cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -815,6 +815,233 @@ class AzureFileSystem::Impl {
+    } else {
+      options.Prefix = internal::EnsureTrailingSlash(base_location.path);
+    }
+    options.PageSizeHint = page_size_hint;
+    options.Include = Azure::Storage::Blobs::Models::ListBlobsIncludeFlags::Metadata;
+
+    // When Prefix.Value() contains a trailing slash and we find a blob that
+    // matches it completely, it is an empty directory marker blob for the
+    // directory we're listing from, and we should skip it.
+    auto is_empty_dir_marker =
+        [&options](const Azure::Storage::Blobs::Models::BlobItem& blob) noexcept -> bool {
+      return options.Prefix.HasValue() && blob.Name == options.Prefix.Value();
+    };
+
+    auto recurse = [&](const std::string& blob_prefix) noexcept -> Status {
+      if (select.recursive && select.max_recursion > 0) {
+        FileSelector sub_select;
+        sub_select.base_dir = base_location.container;
+        sub_select.base_dir += internal::kSep;
+        sub_select.base_dir += internal::RemoveTrailingSlash(blob_prefix);
+        sub_select.allow_not_found = true;
+        sub_select.recursive = true;
+        sub_select.max_recursion = select.max_recursion - 1;
+        return GetFileInfoWithSelectorFromContainer(
+            container_client, context, page_size_hint, sub_select, acc_results);
+      }
+      return Status::OK();
+    };
+
+    // (*acc_results)[*last_dir_reported] is the last FileType::Directory in the results
+    // produced through this loop over the response pages.
+    std::optional<size_t> last_dir_reported{};
+    auto matches_last_dir_reported = [&last_dir_reported,
+                                      acc_results](const FileInfo& info) noexcept {
+      if (!last_dir_reported.has_value() || info.type() != FileType::Directory) {
+        return false;
+      }
+      const auto& last_dir = (*acc_results)[*last_dir_reported];
+      return BasenameView(info.path()) == BasenameView(last_dir.path());
+    };
+
+    auto process_blob =
+        [&](const Azure::Storage::Blobs::Models::BlobItem& blob) noexcept {
+          if (!is_empty_dir_marker(blob)) {
+            const auto& info = acc_results->emplace_back(
+                FileInfoFromBlob(base_location.container, blob));
+            if (info.type() == FileType::Directory) {
+              last_dir_reported = acc_results->size() - 1;
+            }
+          }
+        };
+    auto process_prefix = [&](const std::string& prefix) noexcept -> Status {
+      const std::string path = base_location.container + internal::kSep + prefix;
+      const auto& info = acc_results->emplace_back(DirectoryFileInfoFromPath(path));
+      if (ARROW_PREDICT_FALSE(matches_last_dir_reported(info))) {
+        acc_results->pop_back();
+      } else {
+        last_dir_reported = acc_results->size() - 1;
+        return recurse(prefix);
+      }
+      return Status::OK();
+    };
+
+    try {
+      auto list_response =
+          container_client.ListBlobsByHierarchy(/*delimiter=*/"/", options, context);
```

Comment:
That would require that we build a trie of all the blob names to list them hierarchically. I'm open to doing it, but I think we can delay doing it until there is clear indication that it would help a concrete use-case.